tools/jsonschemavalidator/main.cpp

```diff
@@ -0,0 +1,204 @@
+    throw std::runtime_error(err);
+  }
+  char buffer[4096];
+  FileReadStream fs(fp.get(), buffer, sizeof(buffer));
+  Document doc;
+  doc.ParseStream(fs);
+  if(doc.HasParseError())
+  {
+    throw std::runtime_error("File '" + filePath + "' is not a valid JSON\n Error(offset " + std::to_string(doc.GetErrorOffset()) + "): " + GetParseError_En(doc.GetParseError()));
+  }
+  return doc;
+}
+
+/**
+ * \brief Function to validate json file against a schema file
+ *
+ * \schemafile is the file path of the schema
+ * \jsonfile is the file path of the json data
+ *
+ * \returns std::pair<bool, std::string>. The first element of the pair is true
+ *          if the json file matches the schema, error otherwise. In case of
+ *          error, the second element contains the error report
+ */
+std::pair<bool, std::string> validate(const std::string& jsonfile, const std::string& schemafile) noexcept
+{
+  try
+  {
+    SchemaDocument sd(ReadJSONDocument(schemafile));
+    SchemaValidator validator(sd);
+    Document json = ReadJSONDocument(jsonfile);
+    if (!json.Accept(validator))
+    {
+      std::string errorMsg("JSON file " + jsonfile + " does not match schema file " + schemafile);
+      StringBuffer sb;
+      PrettyWriter<StringBuffer> w(sb);
+      validator.GetError().Accept(w);
+      errorMsg.append("\nError report:\n");
+      errorMsg.append(sb.GetString());
+      throw std::runtime_error(errorMsg);
+    }
+  }
+  catch(std::exception& e)
+  {
+    return std::make_pair(false, e.what());
+  }
+  return std::make_pair(true, "JSON file" + jsonfile + " matches the schema " + schemafile);
+}
+
+/**
+ * Subclass of option::Arg used to validate input options for the program
+ */
+struct Custom_Arg : public option::Arg
+{
+  /**
+   * This callback is used to error out if the option is empty
+   */
+  static option::ArgStatus NonEmpty(const option::Option& option, bool msg)
+  {
+    auto retVal = option::ARG_OK;
+    if (option.arg == 0 || option.arg[0] == 0)
+    {
+      retVal = option::ARG_ILLEGAL;
+      if (msg)
+      {
+        std::cerr << "ERROR: Option '" << option.name << "' requires a non-empty argument" << std::endl;
+      }
+    }
+    return retVal;
+  }
+};
+
+enum OptionIndex
+{
+  UNKNOWN,
+  HELP,
+  SCHEMAFILE,
+  JSONFILE
+};
+
+const option::Descriptor usage[] = {
+  { UNKNOWN,
+    0,
+    "",
+    "",
+    Custom_Arg::None,
+    "\nUSAGE: jsonschemavalidator [options]\n\n"
+    "Options:" },
+  { HELP,
+    0,
+    "h",
+    "help",
+    Custom_Arg::None,
+    " --help, -h  \tPrint usage and exit." },
+  { SCHEMAFILE,
+    0,
+    "s",
+    "schema-file",
+    Custom_Arg::NonEmpty,
+    " --schema-file, -s \tSchema file path" },
+  { JSONFILE,
+    0,
+    "j",
+    "json-file",
+    Custom_Arg::NonEmpty,
+    " --json-file, -j \tJSON file to be validated" },
+  { 0, 0, 0, 0, 0, 0 }
+};
+
+int main(int argc, char *argv[])
+{
+  argc -= (argc > 0);
+  argv += (argc > 0); // skip program name in argv[0]
```

Comment:
Are these two line correct? It looks like it's possible for argc to decrement without argv incrementing. For instance, what if argc is 1? Then the first line sets it to zero, and the second line makes argv NOT increment over the program name.

Reply:
good catch!